alternator/server.cc

```diff
@@ -456,9 +456,16 @@ future<executor::request_return_type> server::handle_api_request(std::unique_ptr
 
     tracing::trace_state_ptr trace_state = maybe_trace_query(client_state, username, op, content);
     tracing::trace(trace_state, "{}", op);
-    rjson::value json_request = co_await _json_parser.parse(std::move(content));
-    co_return co_await callback_it->second(_executor, client_state, trace_state,
-            make_service_permit(std::move(units)), std::move(json_request), std::move(req));
+
+    auto user = client_state.user();
+    auto f = [this, content = std::move(content), &callback = callback_it->second,
+            client_state = std::move(client_state), trace_state = std::move(trace_state),
+            units = std::move(units), req = std::move(req)] () mutable -> future<executor::request_return_type> {
+                rjson::value json_request = co_await _json_parser.parse(std::move(content));
+                co_return co_await callback(_executor, client_state, trace_state,
+                    make_service_permit(std::move(units)), std::move(json_request), std::move(req));
+    };
+    co_return co_await _sl_controller.with_user_service_level(user, std::ref(f));
```

Comment:
Wow, this std::ref thing looks very odd. Why don't you just stuff the entire lambda as a parameter to the function?
Or maybe you just copied the existing code?

By the way, I hope that this with_user_service_level works - and efficiently - in the common (99% of the users) case where this service level feature isn't used. Even more importantly - it must also work (and do nothing) when there is NO user at all - when authentication is disabled.

Reply:
> Or maybe you just copied the existing code?

Yes, this part is taken directly from the enterprise repo (like most of the code in this PR). I believe this particular piece of logic was introduced in https://github.com/scylladb/scylla-enterprise/pull/3218.

I don't see an explanation as to why it was done like this, but given that `with_user_service_level` performs a `std::move` on the reference anyway I think I can safely do the transformation that you proposed.

> By the way, I hope that this with_user_service_level works - and efficiently - in the common (99% of the users) case where this service level feature isn't used.

With service levels on raft this is just a lookup in a `std::map`, so this should be pretty efficient.

> Even more importantly - it must also work (and do nothing) when there is NO user at all - when authentication is disabled.

In that case the lookup is skipped altogether, see the `else` branch in the `with_user_service_level` function.